src/rust/lib/graph-editor/src/component/operator.rs

```diff
@@ -0,0 +1,271 @@
+//! Provides generic operations that can be applied to UI components.
+//!
+//! Rationale: this is a step towards a higher level abstraction for arbitrary UI elements. That is,
+//! instead of doing a lot of low level functionality within our UI components, we should instead be
+//! able to access some data about them and manipulate them in a consistent manner. For example:
+//! instead of every component implementing a "fullscreen mode", they just provide resizing
+//! functionality and access to their su-components and instead of determining the layer of shapes
+//! within the UI component, we instead make the shapes accessible via an API that indicates the
+//! expected behaviour of the shapes, and the actual layer management happens outside of the
+//! component.
+//!
+//! This is also a step to avoid relying too much on the presence of the `Scene` within the UI
+//! component.
+use crate::prelude::*;
+
+use ensogl::display::DomScene;
+use ensogl::display::DomSymbol;
+use ensogl::display::Scene;
+use ensogl::display::Symbol;
+use ensogl::display::scene::View;
+use ensogl::display::traits::*;
+use ensogl::display;
+use ensogl::frp;
+use ensogl::gui::component::animation;
+
+
+
+// ==================================
+// === UI Component Helper Traits ===
+// ==================================
+
+/// Should be implemented by UI component that consist of `Symbol`. Provides access to the shapes
+/// and some helper methods for working with those shapes.
+pub trait NativeComponent {
+    /// Return all `Symbol`s that make up this component.
+    fn symbols(&self) -> Vec<SymbolType>;
+
+    /// Change the scene layer of all `Symbol`s.
+    fn set_layer(&self, layer:&View) {
+        self.symbols().iter().for_each(|symbol| {
+            match symbol{
+                SymbolType::Main(symbol)
+                | SymbolType::Visualisation(symbol)
+                => layer.add(symbol),
+            }
+        })
+    }
+
+    /// Remove the `Symbol`s from all scene layers.
+    fn unset_layers_all(&self, scene:&Scene) {
+        self.symbols().iter().for_each(|symbol|   match symbol{
+            SymbolType::Main(symbol)
+            | SymbolType::Visualisation(symbol)
+            =>scene.views.remove_symbol(symbol),
+        })
+    }
+}
+
+/// Should be implemented by UI component that consist of `DomSymbol`s. Provides access to the
+/// symbols some helper methods for working with them.
+pub trait HtmlComponent {
+    /// Return all `DomSymbol`s that make up this component.
+    fn elements(&self) -> Vec<DomSymbol>;
+
+    /// Change the `DomScene` of all `DomSymbol`s.
+    fn set_layer(&self, scene:&DomScene) {
+        self.elements().iter().for_each(|element| scene.manage(&element))
+    }
+}
```

Comment:
Lets remove it. After we introduce cutting holes to the webgl layer, there will be a single HTML scene, so this will never be used.